src/vision/status_quo/barbara_simulates_hydrodynamics.md

```diff
@@ -0,0 +1,73 @@
+What Barbara wanted to do was find a way to more efficiently use threads: have a fixed number of threads that each mapped to a core on the CPU and assign patches to those threads as patches became ready to compute. The design of the `async` framework seemed to provide exactly that behavior. And to move away from the message passing design, because the number of messages being passed was proportional to the number of trace particles being traced.
+
+As Barbara began working on her new design with `tokio`, her use of `async` went from a general (from the textbook) use of basic `async` features to a more specific implementation leveraging exactly the features that were most suited for her needs.
+
+At first, Barbara was under a false impression about what async executors do. She had assumed that a multi-threaded executor could automatically move the execution of an async block to a worker thread. Then she discovered that async tasks must be explicitly spawned into into a thread pool if they are to be executed on a worker thread. This meant that the algorithm to be parallelized became strongly coupled to both the spawner and the executor. Code that used to cleanly express a physics algorithm now had interspersed references to the task spawner, not only making it harder to understand, but also making it impossible to try different execution strategies, since with Tokio the spawner and executor are the same object (the Tokio runtime). Barbara feels that a better design for data parallelism would enable better separation of concerns: a group of interdependent compute tasks, and a strategy to execute them in parallel.
+
+In order to remove the need for message passing, Barbara moved to a shared state design: she would keep a table tracking the solution state for every grid patch and a specific patch would only start its computation task when solutions were written for all the patches it was dependent on. So, each task needs to access the table with the solution results of all the other tasks. Learning how to properly use shared data with `async` was a new challenge. The initial design:
+
+```rust
+    let mut stage_primitive_and_scalar = |index: BlockIndex, state: BlockState<C>, hydro: H, geometry: GridGeometry| {
+        let stage = async move {
+            let p = state.try_to_primitive(&hydro, &geometry)?;
+            let s = state.scalar_mass / &geometry.cell_volumes / p.map(P::lorentz_factor);
+            Ok::<_, HydroError>( ( p.to_shared(), s.to_shared() ) )
+        };
+        stage_map.insert(index, runtime.spawn(stage).map(|f| f.unwrap()).shared());
+    };
+```
+lacked performance because she needed to clone the value for every task.  So, Barbara switched over to using `Arc` to keep a thread safe RC to the shared data. But this change introduced a lot of `.map` and `.unwrap` function calls, making the code much harder to read. She realized that managing the dependency graph was not intuitive when using `async` for concurrency.
```

Comment:
"she needed to clone _the value_ for every task" - which value are you referring to?

I'm having a hard to relating the code example and the problems mentioned in this paragraph. Perhaps comments in the code examples could bring better attention to the relevant details?

Reply:
Yes, `shared` comes from the `futures` crate. I think the basic pattern I was trying can be represented more cleanly than with the copy-paste's from the code base that are currently in there. I'll propose a cleaner example for the story.

Reply:
To contextualize the snippets in @erichgess's post, here is an approximation of the parallelization strategy in our science codes.

```rust
use std::sync::Arc;
use std::collections::HashMap;
use futures::FutureExt;
use futures::future::join_all;
use tokio::runtime::Runtime;

async fn update(runtime: Runtime, solver: Solver, state: State) -> Result<State, Error> {

    let mut stage1 = HashMap::new();
    let mut stage2 = HashMap::new();
    let mut stage3 = HashMap::new();

    // Stage 1: results can be computed independently on each block. Tokio
    // futures must be 'static, so anything moved into the async block has to
    // be cloned first.
    for (index, block) in &state.patches {
        let block = block.clone();
        let solver = solver.clone();

        let future = async move {
            solver.try_stage1(block)
        };
        stage1.insert(index, runtime.spawn(future).map(|f| f.unwrap()).shared());
    }
    let stage1 = Arc::new(stage1);

    // Stage 2: each result requires multiple blocks froms stage 1.
    for (index, _) in &state.patches {
        let solver = solver.clone();
        let stage1 = stage1.clone();

        let future = async move {
            let neighbors = join_all(solver.neighbors_of(index, stage1)).await;
            solver.try_stage2(index, neighbors)
        };
        stage2.insert(index, runtime.spawn(future).map(|f| f.unwrap()).shared());
    }
    let stage2 = Arc::new(stage2);

    // Stage 3: similar to stage 2
    for (index, _) in &state.patches {

    }
    let stage3 = Arc::new(stage3);

    // Final stage: convert HashMap<Future<Result<T, E>>> -> Future<Result<Vec<T>, E>>
    let new_patches: Result<Vec<_>, _> = join_all(stage3.into_values())
        .await
        .into_iter()
        .collect();

    State { patches: new_patches? }
}

```